pkg/fuzzer/job.go

```diff
@@ -0,0 +1,397 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package fuzzer
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/google/syzkaller/pkg/cover"
+	"github.com/google/syzkaller/pkg/hash"
+	"github.com/google/syzkaller/pkg/ipc"
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/pkg/signal"
+	"github.com/google/syzkaller/prog"
+)
+
+const (
+	smashPrio int64 = iota + 1
+	genPrio
+	triagePrio
+	candidatePrio
+	candidateTriagePrio
+)
+
+type job interface {
+	run(fuzzer *Fuzzer)
+	saveID(id int64)
+	priority() priority
+}
+
+type ProgTypes int
+
+const (
+	ProgCandidate ProgTypes = 1 << iota
+	ProgMinimized
+	ProgSmashed
+	ProgNormal ProgTypes = 0
+)
+
+type jobPriority struct {
+	prio priority
+}
+
+func newJobPriority(base int64) jobPriority {
+	prio := append(make(priority, 0, 2), base)
+	return jobPriority{prio}
+}
+
+func (jp jobPriority) priority() priority {
+	return jp.prio
+}
+
+func (jp *jobPriority) saveID(id int64) {
+	jp.prio = append(jp.prio, id)
```

Comment:
A comment as to why we append id to priority may be useful for future generations.